### PR TITLE
Add e2e test for cronjob chained removal

### PR DESCRIPTION
**Pull request by soltysh**
This is test proving https://github.com/kubernetes/kubernetes/pull/44058 works with cronjobs. This will fail until the aforementioned PR merges. 

@caesarxuchao ptal

### CODE REVIEW

File: test/e2e/workload/cronjob.go [hunk 1]
```diff
@@ -75,6 +75,13 @@ var _ = framework.KubeDescribe("CronJob", func() {
 		By("Removing cronjob")
 		err = deleteCronJob(f.ClientSet, f.Namespace.Name, cronJob.Name)
```
Comment (kargakis) [inline test/e2e/workload/cronjob.go]: deleteCronJob is not setting any DeleteOptions and the default REST strategy for CJs is [orphaning](https://github.com/kubernetes/kubernetes/blob/ee8b663d22b39b515e7e55ca20e25a21aede3ebe/pkg/registry/batch/cronjob/strategy.go#L48) so I am wondering how does this test work?

**Comment by kargakis** [comment]
Unless you are doing a foreground deletion, this test is not going to be stable. Also can you verify that the CronJob is also deleted?

File: test/e2e/workload/cronjob.go (hunk 1 above)
Comment (caesarxuchao) [inline test/e2e/workload/cronjob.go]: Yeah, you need to set the policy to DeletePropagationBackground.

File: test/e2e/workload/cronjob.go [hunk 2]
```diff
@@ -75,6 +75,13 @@ var _ = framework.KubeDescribe("CronJob", func() {
...
 		Expect(err).NotTo(HaveOccurred())
+		By("Ensuring cronjob does not leave jobs nor pods behind")
+		jobs, err = f.ClientSet.Batch().Jobs(f.Namespace.Name).List(metav1.ListOptions{})
```
Comment (caesarxuchao) [inline test/e2e/workload/cronjob.go]: This should be a poll and list. The GC takes time.

File: test/e2e/workload/cronjob.go [hunk 3]
```diff
@@ -75,6 +75,13 @@ var _ = framework.KubeDescribe("CronJob", func() {
...
+		Expect(err).NotTo(HaveOccurred())
+		Expect(jobs.Items).To(HaveLen(0))
+		pods, err := f.ClientSet.Core().Pods(f.Namespace.Name).List(metav1.ListOptions{})
```
Comment (caesarxuchao) [inline test/e2e/workload/cronjob.go]: Same comment here.

**Comment by caesarxuchao** [comment]
@kargakis @soltysh we need to reach consensus on whether to use foreground or background GC.

I think in this test, the stable way is using the foreground gc, then polls for the final deletion of the cronjob, and then verify the pods are deleted.

I think @soltysh's intention was to test the GC properly cascades the background GC. For that, i think it is worth a separate test.

**Comment by kargakis** [comment]
@caesarxuchao should we need to poll in case of foreground deletion? IIUC, the DELETE call should block until everything is deleted, isn't that the case?

**Comment by soltysh** [comment]
@kargakis based on [this](https://github.com/kubernetes/kubernetes/blob/53ac0284736dfe3deeb16bc35d6a23f80c4ee34f/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L365-L369) I'm assuming the blocking depends what we set in ownerRef. [cronjob](https://github.com/kubernetes/kubernetes/blob/6d1da16456966ebf8089f21c0215bdbac952ee42/pkg/controller/cronjob/utils.go#L174) has that set so I'll go with foreground deletion as proposed by @caesarxuchao

**Comment by soltysh** [comment]
@kargakis actually, the foreground removal works without the fix, so we need background removal to verify its role.

**Comment by soltysh** [comment]
@caesarxuchao @kargakis I've created a separate tests just for cronjobs GC, ptal

File: test/e2e/garbage_collector.go [hunk 4]
```diff
@@ -395,8 +431,8 @@ var _ = framework.KubeDescribe("Garbage collector", func() {
 			return len(rsList.Items) > 0, nil
 
 		})
-		if err == wait.ErrWaitTimeout {
-			err = fmt.Errorf("Failed to wait for the Deployment to create some ReplicaSet: %v", err)
+		if err != nil {
```
Comment (kargakis) [inline test/e2e/garbage_collector.go]: You shouldn't overwrite the list error.

Comment (kargakis) [inline test/e2e/garbage_collector.go]: Let me rephrase this one: in case the error is `wait.ErrWaitTimeout`, you need to overwrite that error with the actual cause, that you didn't find any RSs.

Comment (kargakis) [inline test/e2e/garbage_collector.go]: nvmd, this is OK

**Comment by soltysh** [comment]
@caesarxuchao any objections to merge this in?

File: test/e2e/garbage_collector.go [hunk 5]
```diff
@@ -719,4 +755,52 @@ var _ = framework.KubeDescribe("Garbage collector", func() {
...
+				return false, fmt.Errorf("Failed to list jobs: %v", err)
+			}
+			return len(jobs.Items) > 0, nil
+		})
+		if err != nil {
+			framework.Failf("Failed to wait for the CronJob to create some Jobs: %v", err)
+		}
+
+		By("Delete the cronjob")
+		if err := f.ClientSet.BatchV2alpha1().CronJobs(f.Namespace.Name).Delete(cronJob.Name, getNonOrphanOptions()); err != nil {
```
Comment (caesarxuchao) [inline test/e2e/garbage_collector.go]: This could be fixed in another PR. OrphanDependents field is deprecated, set propagationPolicy to [Background](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L364) instead.

File: test/e2e/garbage_collector.go [hunk 6]
```diff
@@ -719,4 +755,52 @@ var _ = framework.KubeDescribe("Garbage collector", func() {
...
+			framework.Failf("Failed to delete the CronJob: %v", err)
+		}
+		By("Wait for the CronJob to be removed")
+		err = wait.PollImmediate(500*time.Millisecond, 1*time.Minute, func() (bool, error) {
+			_, err := f.ClientSet.BatchV2alpha1().CronJobs(f.Namespace.Name).Get(cronJob.Name, metav1.GetOptions{})
+			return err != nil && !errors.IsNotFound(err), nil
+		})
+		By("Verify if cronjob does not leave jobs nor pods behind")
```
Comment (caesarxuchao) [inline test/e2e/garbage_collector.go]: These should be wrapped in a poll. The Cronjob is deleted with getNonOrphanOptions, which is equivalent of the "background" GC policy, which causes the cronjob to be deleted immediately, then GC deletes the dependents in the background.

Comment (caesarxuchao) [inline test/e2e/garbage_collector.go]: If you use the "foreground" GC, then GC will delete the cronjob after all dependents are deleted, then these checks make sense.

**Comment by caesarxuchao** [comment]
@soltysh too late for 1.7, changed to 1.8

**Comment by soltysh** [comment]
@caesarxuchao I've updated per your suggestion and I've also split the fix to rs and deployment test swallowing errors to a separate commit.

**Comment by caesarxuchao** [comment]
/lgtm

@soltysh could you rebase? Thanks.

**Comment by soltysh** [comment]
@caesarxuchao sorry, I was out for almost a month. I've rebased and addressed that single comment of yours. ptal

**Comment by soltysh** [comment]
Bazel fixed, should be good now.

**Comment by soltysh** [comment]
Based on previous approval I'm re-adding the labels.
